conans/test/functional/build_requires/build_requires_test.py

```diff
@@ -350,7 +350,9 @@ class package(ConanFile):
     name            = "second"
     version         = "0.0.0"
     default_options = "first:coverage=True"
-    build_requires  = "first/0.0.0@lasote/stable"
+    
+    def build_requirements(self):
+        self.build_requires("first/0.0.0@lasote/stable", context="host")
```

Comment:
I left a comment in the general PR related to this. If the "install" is not specifying `build` profile or `build` options explicitly, shouldn't the "build" context be considered the "host" one? like a copy. Otherwise, I think we will be breaking many things, not only this usage.

Reply:
It is a little bit different (logic around [conan_api.py#L1187](https://github.com/conan-io/conan/pull/5202/files#diff-633cc9158b73bec9defbb8eb390ba040R1187)):
 * Non-scoped settings/env/options go to `host`. I mean, those without the `:host` or `:build` suffixes, go to host.
 * Then, `build` profile will take explicit values or the `default` profile
 * And `host` profile will take explicit values or be equal to `build` (by default we generate native binaries).

Examples:
 * `conan ... <no args>` will generate binaries using the default profile for `host` and `build`.
 * `conan ... -s os=Linux` will generate binaries for `Linux` (composition rules for profile and arguments apply), using the default profile as `build`.
 * `conan ... -s:h os=Linux`, same as above
 * `conan ... -s:b build_type=Release` will generate binaries for the default profile, but will use `Release` ones for `build` (composition rules for profiles apply).
 * `conan ... -pr:h=host -pr:b=build` will use profile `build` to build and profile `host` for generated artifacts... 

---

The breaking change that I need to revert is that, by default, `build_requires` use the `build` profile and this is breaking for released behavior (but I'm keeping it by now as it helps me to notice about tests that are using profiles and arguments).

Reply:
I need to understand why you think it should behave that way.

The idea I always had is:
   -  The build profile is None by default (if no build profile is specified, that means no explicit value with the `:build` scope, syntax that I like, by the way)
   - If no build profile is available, there is no "scope change" for the build requires.
   - By default a build require scope is the "build" one, but if there is no "build" scope, then there is no change (no breaking)
    - The old `xxx_build` settings keep working exactly the same. While we don't remove it from the settings.yml, they will be present in the build scope when specified (even if it doesn't make any sense)
    - __Everything__ should work exactly the same, no tests changed, nothing, until the user specify any build input.

It is likely that I don't fully understand the reasons behind your design, so please, tell me.

Reply:
ping @jgsogo

Reply:
It is working (or it should work) exactly as you say (taking into account that I will revert the default context for the `build_requires` to be `host` as it is said in the PR).

Also, I'm sure we agree on assigning the same value to `host` and `build` profiles when we are not cross-building so the context change will take place, but it will have no effect. I think it is easier to reason about if we do it this way than adding `if-clauses` in the code.

----

Now, the point we have to talk about, how to get the default values for the profiles, or how to deal with the user input for several scenarios:
 * user provides `host` and `build` profiles, nothing to say, straightforward
 * user provides `host` but not `build` profile:
    + **host** profile will be the one provided by the user
    + **build** profile, we can take the `default` one or make it equal to the `host`:
        - IMO, I would expect Conan to use the `default` profile, as it is usually autodetected and binaries will run in the current machine.
        - If we use the `host` one, we can find that binaries are not compatible with the running machine and build will fail.
 * user provides `build`, but not `host` profile:
    + **build** profile will be the one provided by the user
    + **host** profile can be the `default` one or equal to `build``
        - IMO, we should make it equal to the provided `build` so we are defaulting to a _native compilation_
        - Making it equal to `default` can be also an option, we would be cross-compiling but we would respect the `default` that the user has in the cache.
    To be honest, here I don't know which one is better.
 * user doesn't provide any profile:
    + **build** profile will be equal to `default`
    + **host** profile will be equal to `default` (and equal to `build`)

ping @memsharded 

----

This is important, sure, but I need to know first if the list of features added in this PR are the ones we want to add. If we decide to postpone any of them I need to know asap before starting to modify the PR with your reviews (this PR is merging 6 branches in my repo, I try to keep each sub-feature isolated, but it is hard once you start to modify details and conflicts between branches appear).

Reply:
I already commented on slack about the "sub-features", but I insist it is not relevant until we agree on the approach from my previous comments.

- It is a basic premise of any feature to not change any behavior. That means, if the user doesn't specify a build profile, nothing, not even a comma have to change. Of course, there are exceptions sometimes, but need to be understood and agreed.
- Is not the same to `"revert the default context for the build_requires to be host"` than consider the build profile THE SAME as the host profile when no build profile is specified: The default profile to be applied to a build require is the build profile, but because of the previous point, if no build profile is specified, nothing changes.
- `"user provides host and build profiles"`: In that case the context of the build requires is "build" by default. The possibility of choosing the context of the build requires is also a "sub-feature" that I asked to be in a second iteration. But as the `conanfile.settings_host` and `conanfile.settings_build` sub-feature, if they are already implemented, it makes no sense to rollback.
- `"user provides host but not build profile:"`  In that case, we are not cross-building and everything should work exactly the same (point 2)
    - Conan won't use the default profile for `build` scope.
    - `"If we use the host one, we can find that binaries are not compatible with the running machine and build will fail."`. That's fantastic, I love failures when the input is wrong, exactly the same I currently specify "-s os=Windows" running on Linux.
    - I don't know if internally you prefer to have an exact copy of the host in the build, but indeed we need a way to differentiate when are we activating the "cross build" if we want to change some message. At first sight, I see that as a `None`.
- `"user provides build, but not host profile:"` Specify no host profile means using the default.
- `"user doesn't provide any profile:"` 
    - Host profile will be equal to the default
    - Build profile won't exist (that's why I see it defaulted to None), only an EXPLICIT usage of the build profile will change anything. (point 1)

I don't know the implications of these ideas in the "sub-features, @memsharded and I agreed on this approach even before the kick-off of the feature so we need you to explain if there are additional reasons to not follow the above-described behavior.

Reply:
We agree on every point, especially those related to the stability and not breaking behavior. You can be sure that I've done my best to maintain the existing behavior and to introduce this new feature the best way I've been able to. During the development, I've had to make some decisions that I've shared with you all the first day I had the opportunity to talk about this feature after the kick-off. Nothing has been done here since then waiting for your feedback.

So, about your concerns, I've tried to keep existing behavior the same as it is now (the only change is the one I plan to revert, which is marked as a breaking change since the very beginning. If I did that it is because I think it is very interesting to check how the context switching works in the existing tests).
All the other points we are talking about are new behaviors, some of them are exactly as you are requesting, others are a compromise solution I've realized about during the development, and some of them could be new/unexpected ones because I didn't know everything about this feature when I started it. Nevertheless, when I explained all of this to you in the whiteboard session it looks like we all agree on most of it, nothing has changed since then.
Maybe it is just a matter of the medium, but not the message.

Have you tried to run existing recipes? what about the experience using the new cross-building model? Of course, there will be things broken, but we are talking about a draft.

----

Talking about the implementation of the profiles, these are the top 3 things we have to be sure how to implement them and the implications they have. If you are totally sure about how it should work, it's great, but I would like you to consider the arguments I'm providing as the resulting behavior tries to be exactly the one you described:
 1. Need to know if we want to keep the information about the user input during all the workflow. I mean, if we want to know in every function if the user provided de `build` and `host` profiles explicitly. Current implementation tries to identify asap the `build` and `host` profiles to avoid having to check it afterward. IMO this was the cleaner approach.
 1. Need to know if the default context for a `build_requires` is going to depend on the profiles given by the user. The current implementation does not depend on user input, IMO it is much easier to document and to understand this way.
 1. Need to know how to populate `build` and `host` profile for the 8 possible combinations of the input arguments `--profile`, `--profile:host`, `--profile:build`. Current implementation is making two assumptions:
    + assign the `--profile` argument to the `--profile:host` argument
    + if not `--profile:host` is given (nothing for the `--profile` argument, nothing for the `--profile:host` one), it takes the value of the `--profile:build` so it is defaulting to a _native building_.

I can be wrong about some of these assumptions, but take into account that I'm not trying to _"not follow the above-described behavior"_, I'm just trying to understand it and to provide all the information I can, so we can take the best-informed decision. None of my comments have been about giving a strong opinion about the behavior or how to implement it, and if it has sound like that, I'm so sorry, I was trying just to expose the way it works in this implementation so it's easier for you to understand it when you run some examples using this branch.

Reply:
Summarizing:

You approach:
- The host/build profiles are "calculated" at the beginning and then everything works internally exactly the same. For your comments, it results in a much simpler implementation. You think this would be easier to explain and understand.
- To avoid the breaking behavior, you consider you have to change the default `build_require` context to host.

My initial idea was:
- The explicit `build` profile specification by the user, activates the `build` context. When there is a `build` context available, then the default context for a build requires is the build context. Otherwise, everything keeps exactly the same. In my head, this sounds like the best idea to keep compatibility. I think this doesn't make it harder to explain nor to understand.
- I want to have the "build" context as the default of the `build_requires`, when available.

I'm just reviewing this draft to understand why the approaches are not the same and move this forward in the best direction, so I will dive into the code ASAP to understand better your reasons.

I've just a question that could help to understand, if the default profile for the build requires is the host and I specify a build profile, would the user need to change the recipe to change the default "context" to its build requires?

Reply:
Yes, the summary is right.

Let me add more information related to your question (which is one of the key points of this whole thing):

#### Trying to answer your question...

_(...or to give more information around it. Just repeating myself to have everything in one single comment)_

The recipe declares if the build requires belongs to the `build` context or to the `host` one. For example: `gtest` should belong to the `host` context while a tool like `cmake` belongs to the `build` one. There are no changes depending on user input, it is the recipe writer the one who knows how things should work together (see snippet below). <-- I really think this is important

In my implementation, the user input is always assigned to `host` or to `build` profile, from the very beginning:
 * by default input without scope is assigned to `host` (it is how it works right now, we are compiling binaries to the `host` machine). 
 * if no `build` information is given it will take the `default` profile (usually it is the autodetected one) that is likely to match the `build` profile where we are running the build
 * if no `host` is given it will take the `build` profile, so it will do a _native build_ (not sure if this is the best option, but it is what it is written now)

I would really like to default the `build_requires` to the `build` context because it follows the idea of `build_requires` being tools that we are talking about lately... but it is not possible if we want to keep compatibility 🤷‍♂ 

IMHO, I think it is not a good idea to change the `build_requires` context based on user input, I think that it is the recipe writer the one that should take care of that info:

```python
    def build_requirements(self):
        self.build_requires("cmake/1.0@user/channel", context="build")
        self.build_requires("gtest/1.0@user/channel")  # It is "host" by default (cannot change it)
```

and then the user will get these results:

 * `conan .... --profile=myprofile`: will get binaries for `myprofile` built with tools using `myprofile` too, like it is now.
 * `conan ... --profile:host=phost`: will get binaries for `phost` using the _autodetected profile_ (the default one) for the _tools_.
 * `conan ... --profile:host=phost --profile:build=pbuild`: will get binaries for `phost` using `pbuild` for the _tools_
 * `conan ... --profile:build=pbuilld`: will use `pbuild` for the tools and will default to a native build, so the resulting binaries will use the `pbuild` profile too.

Of course, some of these scenarios are arguable, but this is how it is implemented in this draft.

If you need more info or examples or anything else before having a look at the sources, just let me know, I'm almost writing a book with all of this 😆